main.py

```diff
@@ -1,144 +1,197 @@
+    db: Session = Depends(get_db),
+    role: str = Depends(role_required(["admin", "user"]))
 ):
-    db = SessionLocal()
-
     query = db.query(Pokemon)
 
-    # Filter based on search
+    # Validate keyword datatype based on the search column
+    int_columns = ["total", "hp", "attack", "defense", "sp_atk", "sp_def", "speed", "generation"]
+    bool_columns = ["legendary"]
+    string_columns = ["name", "type_1", "type_2"]
+
     if keyword:
         search_column_attr = getattr(Pokemon, search_column, None)
         if search_column_attr is None:
-            db.close()
             raise HTTPException(status_code=400, detail=f"Invalid search column: {search_column}")
-        query = query.filter(search_column_attr.ilike(f"%{keyword}%"))
+
+        # Validate integer columns
+        if search_column in int_columns:
+            if not keyword.isdigit():
+                raise HTTPException(status_code=400, detail=f"Keyword must be an integer for the {search_column} column")
+            query = query.filter(search_column_attr == int(keyword))
+
+        # Validate boolean columns
+        elif search_column in bool_columns:
+            if keyword.lower() not in ['true', 'false']:
+                raise HTTPException(status_code=400, detail=f"Keyword must be 'true' or 'false' for the {search_column} column")
+            query = query.filter(search_column_attr == (keyword.lower() == 'true'))
+
+        # Convert non-integer inputs to string for string columns
+        elif search_column in string_columns:
+            query = query.filter(search_column_attr.ilike(f"%{keyword}%"))
+        else:
+            raise HTTPException(status_code=400, detail=f"Search not supported for column: {search_column}")
 
     # Sorting
-    if order == "asc":
+    if sort_order == "asc":
         query = query.order_by(asc(sort_by))
     else:
         query = query.order_by(desc(sort_by))
 
-    db_pokemon = query.all()
-    db.close()
+    # Pagination
+    offset = (page - 1) * limit
+    db_pokemon = query.offset(offset).limit(limit).all()
 
     if not db_pokemon:
         raise HTTPException(status_code=404, detail="No Pokémon found")
 
-    return db_pokemon
+    return {"results": db_pokemon}
 
-@app.delete("/pokemon/{number}", response_model=PokemonResponse)
-def delete_pokemon(number: int = Path(description="The number of the Pokémon to delete")):
-    db = SessionLocal()
+
+@app.delete("/pokemon/{number}", response_model=PokemonPostPatchPutOutputSchema)
+def delete_pokemon(number: int = Path(description="The number of the Pokémon to delete"), db: Session = Depends(get_db), role: str = Depends(role_required(["admin", "user"]))):
     db_pokemon = db.query(Pokemon).filter(Pokemon.number == number).first()
     if db_pokemon is None:
-        db.close()
         raise HTTPException(status_code=404, detail="Pokémon not found")
     db.delete(db_pokemon)
     db.commit()
-    db.close()
     return db_pokemon
```

Comment:
We shouldn't return anything in DELETE route
just return the 204 HTTP code if the delete is successful.